File: programmers/Daeseong/programmers_43162_U.cpp
// dfs

#include <string>
#include <vector>

using namespace std;

#define N_MAX 200
bool visited[N_MAX];

void connect(int *n, int target, vector<vector<int>> &computers) {
    visited[target] = true;
    for (int i = 0; i < *n; i++) {
        if (visited[i] || computers[target][i] == 0) continue;
        connect(n, i, computers);
    }
}

int solution(int n, vector<vector<int>> computers) {
    int answer = 0;
    for (int i = 0; i < n; i++) {
        if (!visited[i]) {
            answer += 1;
            connect(&n, i, computers);
        }
    }
    return answer;
}

File: programmers/Daeseong/programmers_43162_U.java
//Solution
//- Using BFS
//- 정석적으로 Queue를 이용했다.
//- visited 배열에서 방문체크를 하며 결과값을 구함.

import java.util.Arrays;
import java.util.LinkedList;
import java.util.Queue;

class Solution {
    public int solution(int n, int[][] computers) {
        int answer = 0;

        Queue<Integer> q = new LinkedList<>();
        boolean[] visited = new boolean[n];
        Arrays.fill(visited, false);

        for(int y = 0; y < n; y++){
            if(visited[y]) continue;
            q.offer(y);
            while (!q.isEmpty()){
                int index = q.poll();
                if(visited[index]) continue;
                visited[index] = true;

                for(int x = 0; x < n; x++){
                    if(visited[x]) continue;
                    if(computers[index][x] == 1) q.offer(x);
                }
            }

            answer++;
        }

        return answer;
    }
}